Add minetest.generate_biomes

Gael-de-Sailly opened the pull request:
## What it does

Implements #8329. Add function `minetest.generate_biomes(vm, pos1, pos2[, noise_filler_depth])` to generate biome nodes in the `VoxelManip` from Lua mapgen mods.

Also updates `heatmap`, `humiditymap` and `biomemap`, so that they can be reused by `minetest.generate_decorations`, `minetest.generate_ores` and `minetest.get_mapgen_object`.

My approach creates a temporary `MapgenBasic` object and calls `MapgenBasic::generateBiomes()`. The class definition is modified to allow such a usecase.

`minetest.generate_biomes` must be called during mapgen (`register_on_generated` callback), and the X/Z extent of `pos1`-`pos2` must match mapgen chunk size, for the buffers to work correctly. I have not found a way to overcome these limitations that would not lead to a much more complex code.

## Why is it interesting/needed?
- Easier to add biomes to Lua mapgens, a single line will replace a long and complex code to re-implement in Lua what already exists in the core (this latter has been done by several modders including me).
- Improves compatibility between Lua mapgens and mods that use the biome system (including biome-dependant decorations/ores). This is critical for a Lua mapgen mod to be interesting for players and server owners. Otherwise it stays an experimental thing that is rarely going to be really used.
- EDIT: Speeds up biome generation compared to Lua-reimplemented biomes (gain is in the range of 50-100 ms). This is also critical for Lua mapgens.

## To do

I consider this PR as Ready for Review.

I initially planned to include a function to generate dust nodes but this may come in a separate PR, this one is already interesting in itself.

## How to test

I have made a [modified version of Paramat's `lvm_example`](https://github.com/Gael-de-Sailly/lvm_example/tree/use_core_biomegen) for testing. It generates very simple terrain, with biomes, ores and decorations. You can append other biome mods like Ethereal to see how it reacts.

![screenshot_20220114_130702](https://user-images.githubusercontent.com/6905002/149513210-e27969c9-2e75-422f-a081-571bcc315204.png)
^ This PR + lvm_example + Ethereal NG

SmallJoker commented on src/mapgen/mapgen.cpp:
```diff
@@ -620,11 +623,22 @@ MapgenBasic::MapgenBasic(int mapgenid, MapgenParams *params, EmergeParams *emerg
+
+	c_stone              = ndef_p->getId("mapgen_stone");
+	c_water_source       = ndef_p->getId("mapgen_water_source");
+	c_river_water_source = ndef_p->getId("mapgen_river_water_source");
```
What about the other class members? Those will contain uninitialized data.

Gael-de-Sailly replied:
They will, but I'm only setting what is actually used by `generateBiomes` (it uses a temporary object).

I set some class members from `l_generate_biomes()` but I need this function for the protected ones.

I can make `ModApiMapgen` a friend class of `MapgenBasic`, leave the definition empty and initialize all class members in `l_generate_biomes` if that's better.

sfan5 replied:
I haven't looked at the code changes yet but _in general_ it is better to refactor the code so that such workarounds are not even necessary.

Gael-de-Sailly replied:
is it better with the new commit?

SmallJoker replied:
(not yet; continued in another comment)

Gael-de-Sailly commented:
I've compared biome generation time between this new function and my Lua re-implementation ([`biomegen` mod](https://gitlab.com/gaelysam/biomegen/)) that closely reproduces the algorithm in `MapgenBasic::generateBiomes()`. The gain of time is around **75 ms** per chunk, which is significant.

Test code: [lvm_example/timer](https://github.com/Gael-de-Sailly/lvm_example/tree/timer), and enable/disable `biomegen` mod to see the difference. Time statistics are displayed at the end.

With the Lua mod:
```
Biomegen calls: 583
Biomegen mean time: 83.899 ms
Biomegen time std: 65.877 ms
```
With `minetest.generate_biomes`:
```
Biomegen calls: 434
Biomegen mean time: 8.565 ms
Biomegen time std: 4.387 ms
```

SmallJoker commented on src/script/lua_api/l_mapgen.cpp:
```diff
@@ -1475,6 +1487,68 @@ int ModApiMapgen::l_generate_decorations(lua_State *L)
+
+	const NodeDefManager *ndef = getServer(L)->getNodeDefManager();
+
+	MapgenBasic mg;
```
Why couldn't you use `mg_current` for this? `MapgenBasic() = default;` does not initialize trivial data types which can quickly result in unpredictable/erroneous code.

How about this instead?
```C++
MapgenBasic *mg = dynamic_cast<MapgenBasic *>(mg_current);
FATAL_ERROR_IF(!mg, "Bug! This mapgen does not inherit from MapgenBasic.");
```
.. and make `MapgenSinglenode` inherit `MapgenBasic` like all other mapgens. If that does not work, you could still specify `virtual` functions in `Mapgen` which the inherited classes would overwrite.

Gael-de-Sailly replied:
I thought redefining `MapgenSinglenode` was out of question, not sure why I was thinking that, maybe paramat was against it a year or so ago. I tried to stay consistent with what was done for `l_generate_ores` and `l_generate_decorations` but I see why it does not make sense to create a temporary object for biomes: it is a strictly mapgen-time method, that ores and decorations are not.

Will do that. `MapgenV6` should also inherit `MapgenBasic` in that case.

sfan5 replied:
Changing existing mapgens to inherit from a class they have no business with just for this does sounds like a bad idea to me.

lhofhansl commented:
Where are we with this? With Lua mapgen being slow (and essentially disabling multi threaded generation due to the env lock) any speed improvement is very welcome!

Zughy commented:
@Gael-de-Sailly are you gonna apply the requested changes?

Zughy commented:
@Rick-W-Storm it means that, even if it's not in the roadmap, some core devs will review it and possibly merge it

Gael-de-Sailly commented:
Rebased. I have in some way rediscovered this PR today (lol), I can start working again on this and I hope this can be reopened. The current version is working but I did not make any change since last commit, except fixups to follow recent changes in minetest.

wsor4035 commented:
![image](https://github.com/minetest/minetest/assets/24964441/dc8293dd-702d-405c-a2e8-65ce254594a1)
cant be reopened, please make a new pr and link to this one @Gael-de-Sailly
